transports/websocket/src/framed.rs

```diff
@@ -0,0 +1,431 @@
+        (Some(Protocol::Dns4(h)), Some(Protocol::Tcp(port))) =>
+            Ok((format!("{}:{}", &h, port), Some(tls::dns_name_ref(&h)?.to_owned()))),
+        (Some(Protocol::Dns6(h)), Some(Protocol::Tcp(port))) =>
+            Ok((format!("{}:{}", &h, port), Some(tls::dns_name_ref(&h)?.to_owned()))),
+        _ => {
+            debug!("multi-address format not supported: {}", addr);
+            Err(Error::InvalidMultiaddr(addr.clone()))
+        }
+    }
+}
+
+// Given a location URL, build a new websocket [`Multiaddr`].
+fn location_to_multiaddr<T>(location: &str) -> Result<Multiaddr, Error<T>> {
+    match Url::parse(location) {
+        Ok(url) => {
+            let mut a = Multiaddr::empty();
+            match url.host() {
+                Some(url::Host::Domain(h)) => {
+                    a.push(Protocol::Dns4(h.into()))
+                }
+                Some(url::Host::Ipv4(ip)) => {
+                    a.push(Protocol::Ip4(ip))
+                }
+                Some(url::Host::Ipv6(ip)) => {
+                    a.push(Protocol::Ip6(ip))
+                }
+                None => return Err(Error::InvalidRedirectLocation)
+            }
+            if let Some(p) = url.port() {
+                a.push(Protocol::Tcp(p))
+            }
+            let s = url.scheme();
+            if s.eq_ignore_ascii_case("https") | s.eq_ignore_ascii_case("wss") {
+                a.push(Protocol::Wss(url.path().into()))
+            } else if s.eq_ignore_ascii_case("http") | s.eq_ignore_ascii_case("ws") {
+                a.push(Protocol::Ws(url.path().into()))
+            } else {
+                debug!("unsupported scheme: {}", s);
+                return Err(Error::InvalidRedirectLocation)
+            }
+            Ok(a)
+        }
+        Err(e) => {
+            debug!("failed to parse url as multi-address: {:?}", e);
+            Err(Error::InvalidRedirectLocation)
+        }
+    }
+}
+
+/// Create a `Connection` from an existing `Framed` value.
+fn new_connection<T, C>(framed: Framed<T, C>, max_size: u64, mode: Mode) -> Connection<T>
+where
+    T: AsyncRead + AsyncWrite
+{
+    let mut codec = base::Codec::new();
+    codec.set_max_data_size(max_size);
+    let old = framed.into_parts();
+    let mut new = FramedParts::new(old.io, codec);
+    new.read_buf = old.read_buf;
+    new.write_buf = old.write_buf;
+    let framed = Framed::from_parts(new);
+    Connection::from_framed(framed, mode)
+}
+
+// BytesConnection ////////////////////////////////////////////////////////////////////////////////
+
+/// A [`Stream`] and [`Sink`] that produces and consumes [`BytesMut`] values
+/// which correspond to the payload data of websocket frames.
+#[derive(Debug)]
+pub struct BytesConnection<T> {
+    inner: Connection<EitherOutput<EitherOutput<client::TlsStream<T>, server::TlsStream<T>>, T>>
+}
+
+impl<T: AsyncRead + AsyncWrite> Stream for BytesConnection<T> {
+    type Item = BytesMut;
+    type Error = io::Error;
```

Comment:
Why is that `io::Error` instead of the error that `self.inner.poll` would return?

Reply:
The inner error is a `soketto::connection::Error` and I wanted to hide the particular websocket implementation we use.

Reply:
Oh, and `RwStreamSink` requires `io::Error` anyway for the `Stream`/`Sink` type it is wrapping.